webhooks: add initial documentation (#365)
gristlabs/grist-help: help/en/docs/webhooks.md

---
description: How to configure webhooks for some external integrations
---

# Webhooks
Webhooks enable you to notify external services whenever rows are
added to a table or if existing rows are modified.

You can configure webhooks from the 'Document Settings' page. Click 'Settings' under the 'Tools' menu found at the bottom of the left-hand navigation panel while viewing a document.

> <center>![webhooks-settings](images/webhooks/settings.png)</center>

Reviewer: The image is a little big now that it's using a 2x version. Could you try shrinking it to half-width? (There's a class you can set - `.screenshot-half` - that appears frequently in the other files.)

Author: Okay, done.

---
description: How to configure webhooks for some external integrations
---

# Webhooks

> 
> [TOC]
> 
> ## Introduction
> 

Reviewer 2: ```suggestion
```

No need for a table of contents. The left-hand nav panel already includes this. We can start with the title `# Webhooks` and then jump into `Webhooks enable you to...`.

Webhooks enable you to notify external services whenever rows are
added to a table or if existing rows are modified.

You can configure webhooks from the document settings at the bottom of the left
sidebar when viewing a document.

> ![webhooks-settings](images/webhooks/settings.png)

Author: Okay, done!

> At the bottom of the settings in the API section, click on the
> "Manage Webhooks" button.

Reviewer 2: ```suggestion
Under the 'API' section of 'Document Settings', click on the
'Manage Webhooks' button.
```

![webhooks-api-settings](images/webhooks/settings-webhooks.png)

This will enable you to define webhooks for your document, where each
row or card in this settings page represents a webhook for your
document. Conceptually, this is similar to the card widget used to

> manage rows to a table.

Reviewer 3: > Conceptually, this is similar to the card widget used to manage rows to a table

Having this here caused me a little confusion as I figured out what it was referring to (admittedly the screenshots don't show in PR review!). I'm not sure if this would be better without this sentence, what do you think?

Author: You're right, it's better without. I've removed that sentence.

   are allowed, but the webhooks will send requests through the
   configured proxy, thus safeguarding your internal Grist endpoints.
   This is the configuration that Grist Labs uses in our cloud-hosted
   environment.

In a trusted environment where malicious users are not expected to
exist, setting `ALLOWED_WEBHOOK_DOMAINS=*` alone may be sufficient
without needing to configure a proxy.

## Payloads

When a webhook is triggered, the rows that matched the webhook's
conditions will generate a JSON array as the webhook's payload. Here
is an example payload.

```json
[
  {
    "id": 29,
    "manualSort": 29,
    "Title": "Trophy",
    "URL": "https://example.com/buy/Trophy",
    "Price": 60,
    "Purchase_status": "wishlisted",
    "Currency": "USD",
    "Play_status": "Not started",
    "Box_art": null,
    "Price_CAD_": 82.362
  },
  {
    "id": 24,
    "manualSort": 24,
    "Title": "Dataman",
    "URL": "https://example.com/buy/Dataman",
    "Price": 50,
    "Purchase_status": "own digitally",
    "Currency": "EUR",
    "Play_status": "Finished",
    "Box_art": null,
    "Price_CAD_": 74.71
  }
]
```

Multiple rows can simultaneously trigger the same webhook. In that

> case, those rows will be sent together in the same payload. The "Ready
> Column" in the webhook's configuration can be useful for gating which

Reviewer 2: ```suggestion
case, those rows will be sent together in the same payload. The 'Ready
Column' in the webhook's configuration can be useful for gating which
```

rows should be sent together.

## Error conditions

If a webhook fails to deliver its payload to the specified URL, it

> will keep retrying periodically. The "Status" column in the webhook

Reviewer 2: ```suggestion
will keep retrying periodically. The 'Status' column in the webhook
```

configuration can be useful for diagnosing any such problems, or to
verify that the payloads were delivered successfully.

### Webhook queue

Webhook payloads are delivered according to a batched queue. Attempts
to reach an endpoint are removed from the queue when a successful
delivery happens. In case there is a problem with the webhook
configuration, it can be helpful to purge the webhook's queue. The

> "clear queue" in the webhook's configuration page will remove any

Reviewer 2: ```suggestion
'Clear Queue' button in the webhook's configuration page will remove any
```
